### DISCUSSION TITLE
Show all found dates as suggestions underneath the date field

### DISCUSSION THREAD

**Discussion by mptr8956** [paperless-ngx/paperless-ngx#384] [Feature Requests] [answered]
I think it would be nice to show all dates that were found in a document as suggested dates right below the date field in the document detail page. It appears very often that there is more than one date on a page. E.g. for an invoice there is usually a date the invoice was created, a date the letter was sent and a date the invoiced service has been performed. There are even documents (car-related, payroll-related, contracts) where a lot more dates can be found on one page.

It would be very helpful to show all dates as suggestions below the date filed (comparable to the suggestions with the correspondents / document types & tags). With a single click on the right date, it would overwrite the first suggestion.

**Comment by sukisoft** [comment]
That would be indeed very useful. But can get very messy i think, some documents have like 5 or more dates on it. But if one can get it to be displayed in a very compact and neat way, why not. Maybe on mobile devices this should be displayed a litte bit different. First thing that comes to my mind would be a "chip" kind of component such as it as with the tags.

**Comment by shamoon** [comment]
I agree :) https://github.com/jonaswinkler/paperless-ng/discussions/187

Edit: I think the idea would be to take the top 2 / 3, display would just be similar to suggestions for the other items, links below the field. But yea, requires backend work, I'm happy to handle front-end, whenever...

**Comment by mptr8956** [comment]
Haha, great. Totally overseen that one in the ng repo. Good to hear that someone else is also having the demand on that. Docspell for example has that functionality and it is super convenient, see for example this screenshot:

https://workpress.plattform32.de/wp-content/uploads/2021/05/Screenshot_2021-02-27-Docspell-scan-2021_01_30_01_03-tiff.png

Is there anyone who is able to implement this feature?

**Comment by hacker-h** [comment]
I would love to see such a feature!
Maybe something like suggesting the first two and last two dates found in the document will cover > 95% of documents

**Reply by aluedtke7** [reply]
I would prefer all dates of the first page

**Reply by skorvek** [reply]
I second the idea of dates on first page- statement and the like will likely have only a couple of dates total. Although with bank statements that can get real messy real quick. So may still need some massaging. Separate thought but more work would be to learn preferred date by correspondent where you likely have the same pattern over and over again.

**Comment by sebastian3107** [comment]
Sometimes it also matches a birthday instead of the actual date of the document. Maybe think about a date blacklist or some limit on how old the matched date can be. I usually scan documents that are <= 2 months old. Anything older would be a false match.

**Reply by hacker-h** [reply]
agree, this could also be a config variable

**Reply by hacker-h** [reply]
also discovered dates in the future could be ignored

**Reply by mptr8956** [reply]
The birthday problem has already been solved with the PAPERLESS_IGNORE_DATES variable! :) But the future dates problem is a good hint as well.

**Reply by sebastian3107** [reply]
where can PAPERLESS_IGNORE_DATES be set in the GUI?

**Reply by mptr8956** [reply]
Not possible in the GUI. If you use docker, it's possible to declare in the environment variables.

**Reply by stumpylog** [reply]
Dates in the future will already be filtered out.  As well as dates older than 1900

**Comment by aluedtke7** [comment]
I could imagine to have all found dates in a dropdown listbox in descending order (newest top, oldest bottom). This should not replace the calendar to manually select a date. It might be enough to only grab the dates of the first page.

**Comment by ajquick** [comment]
I would love to see this implemented in the same way it suggests Correspondents, Document Types and Tags. If there are multiple dates discovered on the document (or if your dd/mm/yyyy changes to mm/dd/yyyy) then it should recommend alternative dates for creation as suggestions. 

I saw comments from another person that the date selected should always be the highest date found, but I don't think that's a good approach either. For example. I have a receipt that shows the purchase date and then shows the date you can return the items by (90 days later). So if it chooses the later date automatically, it would be selecting the wrong date.

Here is what I am looking for as an example:

![image](https://user-images.githubusercontent.com/6588235/170301166-e6818625-7b40-49ea-a84f-0a37e016baaf.png)

Here you can see the date is entered, but it also found a date on the document for the return date.  Making it easy to click one or the other would be a huge help!

**Comment by oscherler** [comment]
I had the same idea for a feature request, so I’m glad to find it here.

I’m willing to tackle the backend work, and I’m glad that @shamoon offered to handle the front-end, as I have marginal experience with Angular.

The difference with the other fields is that their suggestions come form the classifier tables, whereas the date parser directly fills in the date property of the document. I could either:

* start with a temporary proof of concept where I modify the parser slightly to store a selection of suggested dates in an extra column of the document, and we discuss how to do it better if we decide the feature works well-enough to be kept;
* or we could first discuss how to properly store the date options in the DB and where to handle them.

What do you think?

**Reply by tooomm** [reply]
@oscherler There is work happening in #1367 right now in case you want to check it out and chime in!

**Comment by shamoon** [comment, marked as answer]
Happy to report this is implemented in https://github.com/paperless-ngx/paperless-ngx/pull/1367 and will be in the next release. Thanks to [Eckii](https://github.com/Eckii24) for the initial work!
